File: py_libgit/py_libgit/core/head_ref.py
import logging, py_libgit.settings
logger = logging.getLogger(__name__)

import os

class HeadRef:
    def __init__(self):
        logger.info('Create a new HEAD Ref')
        self.pwd = os.getcwd()

    def create_head_ref_file(self, repo_name, bare_repo=False):
        '''Create the HEAD file that holds the hash reference pointing to the top of the working tree

        Keyword arguments:
        repo_name -- the name of the repository
        bare_repo -- specify if this is a bare repo (default False)
        '''
        if bare_repo:
            head_file = os.path.join(self.pwd, repo_name, 'HEAD')
        else:
            head_file = os.path.join(self.pwd, repo_name, '.git', 'HEAD')

        with open(head_file, 'w') as f:
            f.write('ref: refs/heads/master')

    def update_head_ref(self, repo, sha1):
        repo_root = repo.get_repo_root(os.getcwd())
        head_file = os.path.join(repo_root, 'HEAD')
        with open(head_file, 'w') as f:
            f.write(sha1)

    def get_current_head_sha1(self, repo):
        repo_root = repo.get_repo_root(os.getcwd())
        head_file = os.path.join(repo_root, 'HEAD')

        sha1 = '0' * 40
        with open(head_file, 'r') as f:
            content = f.read()

        if content != 'ref: refs/heads/master':
            sha1 = content

        return sha1

File: ruby_libgit/lib/ruby_libgit/head_ref.rb
require 'ruby_libgit'

module RubyLibgit
  # The HEAD Ref class
  class HeadRef
    def initialize
      RubyLibgit::Logging.logger.info('Create a new HEAD Ref')
      @pwd = ::Dir.pwd
    end

    def create_head_ref_file(repo_name, bare_repo = false)
      head_file = if bare_repo
                    ::File.join(@pwd, repo_name, 'HEAD')
                  else
                    ::File.join(@pwd, repo_name, '.git', 'HEAD')
                  end

      ::File.open(head_file, ::File::RDWR | ::File::CREAT, 0o644) do |f|
        f.write('ref: refs/heads/master')
      end
    end
  end
end
